### PR TITLE
Auditing projects for package vulnerabilities during restore

### PR DESCRIPTION
**Pull request by JonDouglas**
Introduction to an auditing experience at package restore time where known vulnerabilities will be reported.

This proposal introduces new MSBuild warnings and output to the NuGet restore operation.

[Rendered Spec](https://github.com/NuGet/Home/blob/dev-feature-vinr/proposed/2022/vulnerabilities-in-restore.md)

Please 👍 or 👎 this comment to help us with the direction of this feature & leave as much feedback/questions/concerns as you'd like on this issue itself and we will get back to you shortly.

Thank You 🎉

### CODE REVIEW

**Comment by mattzink** [comment]
We use Nexus repo manager and Artifactory as local Nuget.org package mirrors, so it would be great if there was a way in the nuget.config file to specify that Nuget.org is used only for vulnerability lookup, not for package hosting.

File: proposed/2022/vulnerabilities-in-restore.md
Comment (loic-sharma) [inline proposed/2022/vulnerabilities-in-restore.md]:
```suggestion
No known vulnerabilities found for <project path>.

```

File: proposed/2022/vulnerabilities-in-restore.md
Comment (loic-sharma) [inline proposed/2022/vulnerabilities-in-restore.md]: What's the value in reporting severities with 0 hits? Would it be better to exclude those instead?

```suggestion
Found 2 vulnerabilities (1 moderate, 1 critical) in 2 package(s)

```

File: proposed/2022/vulnerabilities-in-restore.md
Comment (loic-sharma) [inline proposed/2022/vulnerabilities-in-restore.md:153]: What's the value for this field? Does this really need to be required? Will this be noisy repetitive values?

File: proposed/2022/vulnerabilities-in-restore.md
Comment (loic-sharma) [inline proposed/2022/vulnerabilities-in-restore.md]: Are the `base` and `update` names required if using other partitioning strategies?

Also, what's the value in requiring a maximum of two pages? Does it complicate the client logic to support arbitrary number of pages?

File: proposed/2022/vulnerabilities-in-restore.md
Comment (JonDouglas) [reply proposed/2022/vulnerabilities-in-restore.md]: Good idea, let's keep this open for discussion!

File: proposed/2022/vulnerabilities-in-restore.md
Comment (loic-sharma) [inline proposed/2022/vulnerabilities-in-restore.md]: How big will nuget.org's `base` page be? How quickly will it grow? In the worst case scenario (like an Australian customer), are we concerned about the performance impact of updating this `base` page during a restore?

File: proposed/2022/vulnerabilities-in-restore.md
Comment (nkolev92) [reply proposed/2022/vulnerabilities-in-restore.md]: Base and update aren't required names, but highly recommended.

The idea with 2 pages updating is that the client will always want to have all the data be up to date. 
It would allow refreshing to be super fast without an overly complicated protocol, as we'd only download the out of date files.

It's balance between CI (where the data is always download) and local, where the data is refreshed.

File: proposed/2022/vulnerabilities-in-restore.md
Comment (loic-sharma) [inline proposed/2022/vulnerabilities-in-restore.md]: Will vulnerability reporting take source maps into account? If two package sources report different vulnerabilities for the same package, will the client report all vulnerabilities?

File: proposed/2022/vulnerabilities-in-restore.md
Comment (nkolev92) [reply proposed/2022/vulnerabilities-in-restore.md]: These were the numbers a few weeks ago:

vulnerability_all_alternate.json    2.4 Mb
vulnerability_all_alternate.json.gz 78.9 Kb

> are we concerned about the performance impact of updating this base page during a restore?

Definitely a consideration. We'd only redownload when restore actually happens. The whole process takes time, but right now we think those numbers are relatively small and a cost worth paying given the benefit.

File: proposed/2022/vulnerabilities-in-restore.md
Comment (loic-sharma) [inline proposed/2022/vulnerabilities-in-restore.md:263]: Would it make sense to update vulnerability info in the background in Visual Studio? That way, restore performance in Visual Studio wouldn't be affected by this feature.

File: proposed/2022/vulnerabilities-in-restore.md
Comment (loic-sharma) [inline proposed/2022/vulnerabilities-in-restore.md:404]: Perhaps there could be a nuget.config for a list of vulnerability info URLs? That'd be useful if I have a single private package source that doesn't point to its upstream vulnerability info URL yet.

File: proposed/2022/vulnerabilities-in-restore.md
Comment (loic-sharma) [reply proposed/2022/vulnerabilities-in-restore.md]: The 2 page idea makes perfect sense. I'm more curious as to why we *require* a maximum of 2 pages; that seems odd.

Comment (nkolev92) [reply proposed/2022/vulnerabilities-in-restore.md]: It's not really suggesting maximum of 2 pages.
Or at least, it's not supposed to :) 

What that line is trying to communicate is that as few pages as possible should be updated, so that the client is not forced to redownload pages all the time.

File: proposed/2022/vulnerabilities-in-restore.md
Comment (kapsiR) [reply proposed/2022/vulnerabilities-in-restore.md:263]: Good idea!  
Additionally this could be another step in CI (cli), so that not every build would be affected by a new download or in other words: 
Have more control over the time when a new file is going to be downloaded

File: proposed/2022/vulnerabilities-in-restore.md
Comment (kapsiR) [inline proposed/2022/vulnerabilities-in-restore.md]: I don't know exactly what sources this database is using but are there any plans to support additional databases too? (e.g. Google osv.dev)

File: proposed/2022/vulnerabilities-in-restore.md
Comment (nkolev92) [reply proposed/2022/vulnerabilities-in-restore.md:263]: > Would it make sense to update vulnerability info in the background in Visual Studio? That way, restore performance in Visual Studio wouldn't be affected by this feature.

We don't expect this to be taking so long where it's noticeable consistently. 

On the topic of background downloading, it's not always desirable to communicate with sources without user intent. That could for example trigger authentication and we wouldn't want anything to interrupt the user. 

> Additionally this could be another step in CI (cli), so that not every build would be affected by a new download or in other words:

There are a lot of optimizations that happen prior to vulnerability data being downloaded. 
If you're not changing your packages/deleting the obj folder, restore will be snappy and won't redownload sources.

File: proposed/2022/vulnerabilities-in-restore.md
Comment (nkolev92) [reply proposed/2022/vulnerabilities-in-restore.md:404]: Yes, it would have to be something along those lines.

File: proposed/2022/vulnerabilities-in-restore.md
Comment (nkolev92) [reply proposed/2022/vulnerabilities-in-restore.md]: That's a great question. 

Right now, we think no is the answer for that, we "merge" the data, but we can certainly change that based on community input.

File: proposed/2022/vulnerabilities-in-restore.md
Comment (zivkan) [reply proposed/2022/vulnerabilities-in-restore.md]: The current proposal is that NuGet client gets the list of vulnerabilities from the NuGet package source. How the server implements the vulnerability information resource is up to the server implementation. They certainly could use osv.dev.

Looking at osv.dev, I don't see a way to download the entire database (unless I need to clone, build, and run the github repo?). So, from a NuGet client point of view, osv.dev doesn't fit well with our design.

This could be something added in the future if there's enough demand, but the current plan is that NuGet server implementations are responsible for providing the client with the relevant data.

Comment (kapsiR) [reply proposed/2022/vulnerabilities-in-restore.md]: Thanks to clarify on that, maybe it's worth adding to the docs? 

FYI there are data dumps available:
https://github.com/google/osv.dev#data-dumps

Comment (kapsiR) [reply proposed/2022/vulnerabilities-in-restore.md]: And more important: GH advisory db is a source of osv.dev! 😉

**Comment by CamiloTerevinto** [comment]
This sounds very interesting, thanks @JonDouglas. I have a few comments:

* CLI switches to run/skip auditing 

It's quite common for an entire build to be run twice (feature branch and dev/main branch) for a change in environments with CI/CD. I think that for this to be useful, we should have an easy way of telling `dotnet restore` to run or skip the audit (for example, run the audit on the PR build, skip it in the dev build).

* Fail on X level

It would be useful to have a way to make `dotnet restore` exit as failure when a package with a particular vulnerability level is found (such as treating `NU1904` as error instead of warning).  
However, this might be doable already through something like `<WarningsAsErrors>NU1904</WarningsAsErrors>`.

File: proposed/2022/vulnerabilities-in-restore.md
Comment (zivkan) [reply proposed/2022/vulnerabilities-in-restore.md]: FWIW, I tested a different file format for the vulnerability data, trading off computation speed vs file size, with the idea that smaller file size will parse more quickly, so the slower per-package computation might still come ahead. Benchmarks suggest yes, overall perf is better, and the current file size for all github reviewed vulnerabilities is 175KB uncompressed, 9KB gzipped. So, we'll be going ahead with the smaller file size format instead now.

**Comment by nkolev92** [comment]
> CLI switches to run/skip auditing

Any CLI switch basically doubles as an MSBuild property. We're proposing MSBuild properties for opting into auditing, so that automatically makes it available as a CLI switch. Example: `dotnet restore -p:NuGetAudit=disable`

File: proposed/2022/vulnerabilities-in-restore.md
Comment (zivkan) [inline proposed/2022/vulnerabilities-in-restore.md:371]: @JonDouglas FYI, I moved the vulnerability summary output on command line restores under "Future Possibilities" for two reasons. 1. So we can get a V1 of this feature out sooner. 2. I don't believe this proposal has been specced out well. I have a bunch of questions if we want to implement it.

For example, in the sample output, if there are non-vulerability warnings or errors, where in the output do they go?  Before the vulnerability summary, or below with the sample vulerability warnings?   If before, should vulerability warnings/errors be output twice, once between the summary once after?

What's the expected order of output when restoring a solution with many projects? Wait until all projects finish restoring, and then output all their summaries?  Or output project summaries as each project finishes restoring, even if it means that the vulnerability summary for that project will be somewhere much earlier in the restore log?

When restoring a solution where the same package version is in multiple projects, should we de-duplicate the output?

Anyway, as I said, I think we can take this on in the future if we still think it's useful. Let's get a V1 of the feature done without it, and see how we and customers feel about it.

File: proposed/2022/vulnerabilities-in-restore.md
Comment (zivkan) [reply proposed/2022/vulnerabilities-in-restore.md]: I updated the spec to try to make it clearer that this is our suggestion, but servers can implement the data partitioning in different ways. The client will just look at the list of files, and their update timestamps to determine whether it can use a previously cached download or not.

File: proposed/2022/vulnerabilities-in-restore.md
Comment (zivkan) [reply proposed/2022/vulnerabilities-in-restore.md]: I updated the spec to try to make this clearer.

File: proposed/2022/vulnerabilities-in-restore.md
Comment (nkolev92) [inline proposed/2022/vulnerabilities-in-restore.md:63]: If you put bash, you get better highlighting :) Not a big deal :D

File: proposed/2022/vulnerabilities-in-restore.md
Comment (nkolev92) [inline proposed/2022/vulnerabilities-in-restore.md]: maybe mention that it includes the prerelease versions in the range as well?

File: proposed/2022/vulnerabilities-in-restore.md
Comment (nkolev92) [reply proposed/2022/vulnerabilities-in-restore.md:371]: I think some details, like the ones about the handling of multiple vulnerability for a package can still be in the main part as that's likely going to be done.

File: proposed/2022/vulnerabilities-in-restore.md
Comment (nkolev92) [inline proposed/2022/vulnerabilities-in-restore.md]: This can probably be in the main part.

**Review by nkolev92 — COMMENTED** [review]
Great updates @zivkan

File: proposed/2022/vulnerabilities-in-restore.md
Comment (JonDouglas) [reply proposed/2022/vulnerabilities-in-restore.md:371]: Let's see what this shapes to be while working to answer the unknown. Thanks for being so thorough with questions to be answered.

File: proposed/2022/vulnerabilities-in-restore.md
Comment (zivkan) [reply proposed/2022/vulnerabilities-in-restore.md:63]: The only highlighting I see are the words `in`, `for` and the `<` and `>` characters. Which make sense in bash scripts, but in my opinion it doesn't make sense for these samples.  Why in the sentence "Found 2 vulerabilities in 1 page(s)" should only the word `in`, be highlighted?  🤷

Comment (nkolev92) [reply proposed/2022/vulnerabilities-in-restore.md:63]: Here's what I am seeing: 

image

The quoted part is being highlighted. It's not a big deal either way, and most definitely a nit :) Your call.

**Review by nkolev92 — APPROVED** [review]

**Comment by JonDouglas** [comment]
We have approval from @nkolev92, but @zivkan did you want to approve before merge?

File: proposed/2022/vulnerabilities-in-restore.md
Comment (zivkan) [inline proposed/2022/vulnerabilities-in-restore.md]:
```suggestion
    "@type": "VulnerabilityInfo/6.6.0",
```

Oops, I already changed this elsewhere in the doc, but I missed it here.

**Review by zivkan — APPROVED** [review]

**Pull request merged**
